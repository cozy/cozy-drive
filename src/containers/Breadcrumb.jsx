import styles from '../styles/breadcrumb'

import { ROOT_DIR_ID, TRASH_DIR_ID } from '../constants/config'

import React from 'react'
import withState from 'cozy-ui/react/helpers/withState'
import { translate } from '../lib/I18n'
import { Link, withRouter } from 'react-router'
import { connect } from 'react-redux'
import { openFolder } from '../actions'
import classNames from 'classnames'
import Spinner from '../components/Spinner'

import { getFolderPath, getFolderUrl } from '../reducers'

const Breadcrumb = ({ t, router, location, path, opening, deployed, toggleOpening, toggleDeploy, goToFolder }) => {
  if (!path) {
    return null
  }

  path.forEach(folder => {
    if (folder.id === ROOT_DIR_ID) {
      folder.name = t('breadcrumb.title_files')
    } else if (folder.id === TRASH_DIR_ID) {
      folder.name = t('breadcrumb.title_trash')
    }
    if (!folder.name) folder.name = 'â€¦'
  })

  return (
    <div
      className={classNames(styles['fil-path-backdrop'], {[styles['deployed']]: deployed})}
      onClick={toggleDeploy}
    >
      {path.length >= 2 &&
        <Link
          to={getFolderUrl(path[path.length - 2].id, location)}
          className={styles['fil-path-previous']}
          onClick={e => {
            e.preventDefault()
            toggleOpening()
            if (deployed) toggleDeploy()
            goToFolder(path[path.length - 2].id).then(() => {
              toggleOpening()
              router.push(getFolderUrl(path[path.length - 2].id, location))
            })
          }}
        />
      }
      <h2 className={styles['fil-path-title']}>

        { path.map((folder, index) => {
          if (index < path.length - 1) {
            return <Link
              to={getFolderUrl(folder.id, location)}
              className={styles['fil-path-link']}
              onClick={e => {
                e.preventDefault()
                toggleOpening()
                if (deployed) toggleDeploy()
                goToFolder(folder.id).then(() => {
                  toggleOpening()
                  router.push(getFolderUrl(folder.id, location))
                })
              }}
            >
              <a>
                { folder.name }
              </a>
              <span className={styles['fil-path-separator']}>/</span>
            </Link>
          } else {
            return <span
              className={styles['fil-path-current']}
              onClick={e => {
                e.stopPropagation()
                if (path.length >= 2) toggleDeploy()
              }}
            >
              { folder.name }
              {path.length >= 2 &&
                <span className={styles['fil-path-down']} />
              }

              { opening && <Spinner /> }
            </span>
          }
        }) }

      </h2>
    </div>
  )
}

const mapStateToProps = (state, ownProps) => ({
  path: getFolderPath(state, ownProps.location)
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  goToFolder: folderId => dispatch(openFolder(folderId))
})

export default withRouter(translate()(connect(
  mapStateToProps,
  mapDispatchToProps
)(withState({
  opening: false,
  deployed: false
}, (setState) => ({
  toggleOpening: () => {
    setState(state => ({ opening: !state.opening }))
  },
  toggleDeploy: () => {
    setState(state => ({ deployed: !state.deployed }))
  }
}))(Breadcrumb))))
