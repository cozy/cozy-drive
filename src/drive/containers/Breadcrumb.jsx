/* global __TARGET__ cozy */
import styles from '../styles/breadcrumb'

import { ROOT_DIR_ID, TRASH_DIR_ID } from '../constants/config'

import React, { Component } from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { Link, withRouter } from 'react-router'
import { connect } from 'react-redux'
import { openFolder } from '../actions'
import classNames from 'classnames'
import Spinner from 'cozy-ui/react/Spinner'
import { withBreakpoints } from 'cozy-ui/react'

import { getFolderPath, getFolderUrl } from '../reducers'

const { BarCenter, BarLeft } = cozy.bar

const renamePathNames = (path, location, t) => {
  if (location.pathname === '/recent') {
    path.unshift({
      name: t('breadcrumb.title_recent')
    })
  }

  path.forEach(folder => {
    if (folder.id === ROOT_DIR_ID) {
      folder.name = t('breadcrumb.title_drive')
    } else if (folder.id === TRASH_DIR_ID) {
      folder.name = t('breadcrumb.title_trash')
    }
    if (!folder.name) folder.name = 'â€¦'
  })

  return path
}

class Breadcrumb extends Component {
  state = {
    opening: false,
    deployed: false
  }

  toggleOpening = () => {
    this.setState(state => ({ opening: !state.opening }))
  }

  toggleDeploy = () => {
    this.state.deployed ? this.closeMenu() : this.openMenu()
  }

  openMenu() {
    this.setState({ deployed: true })
    document.addEventListener('click', this.handleClickOutside, true)
  }

  closeMenu() {
    this.setState({ deployed: false })
    document.removeEventListener('click', this.handleClickOutside, true)
  }

  navigateToFolder = (e, folderId) => {
    const { router, location, goToFolder, getFolderUrl } = this.props
    e.preventDefault()

    this.toggleOpening()
    if (this.state.deployed) this.toggleDeploy()

    goToFolder(folderId).then(() => {
      this.toggleOpening()
      this.toggleDeploy()
      router.push(getFolderUrl(folderId, location))
    })
  }

  handleClickOutside = e => {
    if (!this.menu.contains(e.target)) {
      e.stopPropagation()
      this.closeMenu()
    }
  }

  render() {
    const { location, path } = this.props
    const { opening, deployed } = this.state

    return path ? (
      <div
        className={classNames(
          styles['fil-path-backdrop'],
          { [styles['deployed']]: deployed },
          { [styles['mobile']]: __TARGET__ === 'mobile' }
        )}
      >
        <h2
          className={styles['fil-path-title']}
          onClick={this.toggleDeploy}
          ref={ref => {
            this.menu = ref
          }}
        >
          {path.map((folder, index) => {
            if (index < path.length - 1) {
              return (
                <Link
                  to={getFolderUrl(folder.id, location)}
                  className={styles['fil-path-link']}
                  onClick={e => this.navigateToFolder(e, folder.id)}
                >
                  <a>{folder.name}</a>
                  <span className={styles['fil-path-separator']}>/</span>
                </Link>
              )
            } else {
              return (
                <span
                  className={styles['fil-path-current']}
                  onClick={e => {
                    e.stopPropagation()
                    if (path.length >= 2) this.toggleDeploy()
                  }}
                >
                  <span className={styles['fil-path-current-name']}>
                    {folder.name}
                  </span>
                  {path.length >= 2 && (
                    <span className={styles['fil-path-down']} />
                  )}

                  {opening && <Spinner />}
                </span>
              )
            }
          })}
        </h2>
      </div>
    ) : null
  }
}

class PreviousButton extends Component {
  navigateToFolder = folderId => {
    const { router, location, goToFolder, getFolderUrl } = this.props
    goToFolder(folderId).then(() => {
      router.push(getFolderUrl(folderId, location))
    })
  }

  render({ location, path }) {
    const previousFolderId = path[path.length - 2].id
    return (
      <button
        className={styles['fil-path-previous']}
        onClick={() => this.navigateToFolder(previousFolderId)}
      />
    )
  }
}

const MobileAwareBreadcrumb = props => {
  return props.breakpoints.isMobile ? (
    <div>
      {props.path.length >= 2 && (
        <BarLeft>
          <PreviousButton {...props} />
        </BarLeft>
      )}
      <BarCenter>
        <Breadcrumb {...props} />
      </BarCenter>
    </div>
  ) : (
    <Breadcrumb {...props} />
  )
}

const mapStateToProps = (state, ownProps) => ({
  path: renamePathNames(
    getFolderPath(state, ownProps.location, ownProps.isPublic),
    ownProps.location,
    ownProps.t
  ),
  getFolderUrl
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  goToFolder: folderId => dispatch(openFolder(folderId))
})

export default withBreakpoints()(
  withRouter(
    translate()(
      connect(mapStateToProps, mapDispatchToProps)(MobileAwareBreadcrumb)
    )
  )
)
