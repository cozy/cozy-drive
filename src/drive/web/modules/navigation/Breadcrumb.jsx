/* global __TARGET__ cozy */
import React, { Component } from 'react'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'
import classNames from 'classnames'
import Spinner from 'cozy-ui/react/Spinner'
import Icon from 'cozy-ui/react/Icon'
import { translate } from 'cozy-ui/react/I18n'
import { withBreakpoints } from 'cozy-ui/react'
import SharedDocuments from 'sharing/components/SharedDocuments'

import { ROOT_DIR_ID, TRASH_DIR_ID } from 'drive/constants/config'
import { openFolder, getFolderUrl } from 'drive/web/modules/navigation/duck'
import getFolderPath from './getFolderPath'

import styles from './breadcrumb.styl'

export const renamePathNames = (path, pathname, t) => {
  if (pathname === '/recent') {
    path.unshift({
      name: t('breadcrumb.title_recent')
    })
  } else if (pathname.match(/^\/sharings/)) {
    path.unshift({
      name: t('breadcrumb.title_sharings'),
      url: '/sharings'
    })
  }

  path.forEach(folder => {
    if (folder.id === ROOT_DIR_ID) {
      folder.name = t('breadcrumb.title_drive')
    } else if (folder.id === TRASH_DIR_ID) {
      folder.name = t('breadcrumb.title_trash')
    }
    if (!folder.name) folder.name = 'â€¦'
  })

  return path
}

export class Breadcrumb extends Component {
  state = {
    deployed: false
  }

  toggleDeploy = () => {
    this.state.deployed ? this.closeMenu() : this.openMenu()
  }

  openMenu() {
    this.setState({ deployed: true })
    document.addEventListener('click', this.handleClickOutside, true)
  }

  closeMenu() {
    this.setState({ deployed: false })
    document.removeEventListener('click', this.handleClickOutside, true)
  }

  handleClickOutside = e => {
    if (this.menu && !this.menu.contains(e.target)) {
      e.stopPropagation()
      this.closeMenu()
    }
  }

  render() {
    const {
      path,
      onBreadcrumbClick,
      opening,
      inlined,
      className = ''
    } = this.props
    const { deployed } = this.state

    if (!path) return false

    return (
      <div
        className={classNames(
          styles['fil-path-backdrop'],
          { [styles['deployed']]: deployed },
          { [styles['inlined']]: inlined },
          { [styles['mobile']]: __TARGET__ === 'mobile' },
          className
        )}
      >
        <h2
          data-test-id="path-title"
          className={styles['fil-path-title']}
          onClick={this.toggleDeploy}
          ref={ref => {
            this.menu = ref
          }}
        >
          {path.map((folder, index) => {
            if (index < path.length - 1) {
              return (
                <span
                  className={styles['fil-path-link']}
                  onClick={e => {
                    e.stopPropagation()
                    onBreadcrumbClick(folder)
                  }}
                  key={index}
                >
                  <span className={styles['fil-path-link-name']}>
                    {folder.name}
                  </span>
                  <Icon
                    icon="forward"
                    className={styles['fil-path-separator']}
                  />
                </span>
              )
            } else {
              return (
                <span
                  className={styles['fil-path-current']}
                  onClick={e => {
                    e.stopPropagation()
                    if (path.length >= 2) this.toggleDeploy()
                  }}
                  key={index}
                >
                  <span className={styles['fil-path-current-name']}>
                    {folder.name}
                  </span>
                  {path.length >= 2 && (
                    <span className={styles['fil-path-down']} />
                  )}

                  {opening && <Spinner />}
                </span>
              )
            }
          })}
        </h2>
      </div>
    )
  }
}

class RouterBreadCrumb extends Component {
  state = {
    opening: false
  }

  toggleOpening = () => {
    this.setState(state => ({ opening: !state.opening }))
  }

  navigateToFolder = folderId => {
    const { router, location, goToFolder, getFolderUrl } = this.props

    this.toggleOpening()

    goToFolder(folderId).then(() => {
      this.toggleOpening()
      router.push(getFolderUrl(folderId, location))
    })
  }

  navigateToPath = path => {
    const { router } = this.props
    router.push(path)
  }

  navigateTo = folder => {
    if (folder.id) this.navigateToFolder(folder.id)
    else this.navigateToPath(folder.url)
  }

  render() {
    const { path } = this.props
    const { opening } = this.state

    return (
      <Breadcrumb
        path={path}
        onBreadcrumbClick={this.navigateTo}
        opening={opening}
      />
    )
  }
}

export const PreviousButton = ({ onClick }) => (
  <button className={styles['fil-path-previous']} onClick={onClick} />
)

class RouterPreviousButton extends Component {
  navigateToFolder = folderId => {
    const { router, location, goToFolder, getFolderUrl } = this.props
    goToFolder(folderId).then(() => {
      router.push(getFolderUrl(folderId, location))
    })
  }

  navigateToPath = path => {
    const { router } = this.props
    router.push(path)
  }

  navigateBack = () => {
    const previousSegment = this.props.path[this.props.path.length - 2]
    if (previousSegment.id) this.navigateToFolder(previousSegment.id)
    else this.navigateToPath(previousSegment.url)
  }

  render() {
    return <PreviousButton onClick={this.navigateBack} />
  }
}

const MobileAwareBreadcrumb = props => {
  const { BarCenter, BarLeft } = cozy.bar

  return props.breakpoints.isMobile ? (
    <div>
      {props.path.length >= 2 && (
        <BarLeft>
          <RouterPreviousButton {...props} />
        </BarLeft>
      )}
      <BarCenter>
        <RouterBreadCrumb {...props} />
      </BarCenter>
    </div>
  ) : (
    <RouterBreadCrumb {...props} />
  )
}

const mapStateToProps = (state, ownProps) => ({
  path: renamePathNames(
    getFolderPath(
      state.view.displayedFolder,
      ownProps.location.pathname,
      ownProps.isPublic,
      ownProps.sharedDocuments
    ),
    ownProps.location.pathname,
    ownProps.t
  ),
  getFolderUrl
})

const mapDispatchToProps = dispatch => ({
  goToFolder: folderId => dispatch(openFolder(folderId))
})

const withSharedDocuments = Wrapped =>
  class withSharedDocumentsClass extends React.Component {
    render() {
      return (
        <SharedDocuments>
          {({ sharedDocuments }) => (
            <Wrapped sharedDocuments={sharedDocuments} {...this.props} />
          )}
        </SharedDocuments>
      )
    }
  }

export default withBreakpoints()(
  withRouter(
    translate()(
      withSharedDocuments(
        connect(
          mapStateToProps,
          mapDispatchToProps
        )(MobileAwareBreadcrumb)
      )
    )
  )
)
