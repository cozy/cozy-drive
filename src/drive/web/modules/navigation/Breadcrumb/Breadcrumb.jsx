/* global __TARGET__ */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Icon from 'cozy-ui/transpiled/react/Icon'

import styles from 'drive/web/modules/navigation/Breadcrumb/breadcrumb.styl'

export class Breadcrumb extends Component {
  state = {
    deployed: false
  }

  toggleDeploy = () => {
    this.state.deployed ? this.closeMenu() : this.openMenu()
  }

  openMenu() {
    this.setState({ deployed: true })
    document.addEventListener('click', this.handleClickOutside, true)
  }

  closeMenu() {
    this.setState({ deployed: false })
    document.removeEventListener('click', this.handleClickOutside, true)
  }

  handleClickOutside = e => {
    if (this.menu && !this.menu.contains(e.target)) {
      e.stopPropagation()
      this.closeMenu()
    }
  }

  render() {
    const {
      path,
      onBreadcrumbClick,
      opening,
      inlined,
      className = ''
    } = this.props
    const { deployed } = this.state

    if (!path) return false

    return (
      <div
        className={cx(
          styles['fil-path-backdrop'],
          { [styles['deployed']]: deployed },
          { [styles['inlined']]: inlined },
          { [styles['mobile']]: __TARGET__ === 'mobile' },
          className
        )}
      >
        <h2
          data-test-id="path-title"
          className={styles['fil-path-title']}
          onClick={this.toggleDeploy}
          ref={ref => {
            this.menu = ref
          }}
        >
          {path.map((folder, index) => {
            if (index < path.length - 1) {
              return (
                <span
                  className={styles['fil-path-link']}
                  onClick={e => {
                    e.stopPropagation()
                    onBreadcrumbClick(folder)
                  }}
                  key={index}
                >
                  <span className={styles['fil-path-link-name']}>
                    {folder.name}
                  </span>
                  <Icon icon="right" className={styles['fil-path-separator']} />
                </span>
              )
            } else {
              return (
                <span
                  className={styles['fil-path-current']}
                  onClick={e => {
                    e.stopPropagation()
                    if (path.length >= 2) this.toggleDeploy()
                  }}
                  key={index}
                >
                  <span className={styles['fil-path-current-name']}>
                    {folder.name}
                  </span>
                  {path.length >= 2 && (
                    <span className={styles['fil-path-down']} />
                  )}

                  {opening && <Spinner />}
                </span>
              )
            }
          })}
        </h2>
      </div>
    )
  }
}

Breadcrumb.propTypes = {
  path: PropTypes.array,
  onBreadcrumbClick: PropTypes.func,
  opening: PropTypes.bool,
  inlined: PropTypes.bool,
  className: PropTypes.string
}

export default Breadcrumb
