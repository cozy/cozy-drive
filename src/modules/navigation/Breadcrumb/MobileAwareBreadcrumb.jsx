import React from 'react'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import MobileBreadcrumb from './components/MobileBreadcrumb/MobileBreadcrumb'
import DesktopBreadcrumb from './components/DesktopBreadcrumb/DesktopBreadcrumb'

export const MobileAwareBreadcrumb = props => {
  const { isMobile } = useBreakpoints()

  return isMobile ? (
    <MobileBreadcrumb {...props} />
  ) : (
    <DesktopBreadcrumb {...props} />
  )
}

export default MobileAwareBreadcrumb
