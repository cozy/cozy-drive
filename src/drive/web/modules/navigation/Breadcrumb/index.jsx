import React from 'react'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'

import { translate } from 'cozy-ui/transpiled/react/I18n'
import { withBreakpoints } from 'cozy-ui/transpiled/react'
import SharedDocuments from 'cozy-sharing/dist/components/SharedDocuments'

import { ROOT_DIR_ID, TRASH_DIR_ID } from 'drive/constants/config'
import { openFolder, getFolderUrl } from 'drive/web/modules/navigation/duck'
import getFolderPath from './getFolderPath'
import MobileAwareBreadcrumb from './MobileAwareBreadcrumb'

import Breadcrumb from './Breadcrumb'
import PreviousButton from './PreviousButton'

export { Breadcrumb, PreviousButton }

export const renamePathNames = (path, pathname, t) => {
  if (pathname === '/recent') {
    path.unshift({
      name: t('breadcrumb.title_recent')
    })
  } else if (pathname.match(/^\/sharings/)) {
    path.unshift({
      name: t('breadcrumb.title_sharings'),
      url: '/sharings'
    })
  }

  path.forEach(folder => {
    if (folder.id === ROOT_DIR_ID) {
      folder.name = t('breadcrumb.title_drive')
    } else if (folder.id === TRASH_DIR_ID) {
      folder.name = t('breadcrumb.title_trash')
    }
    if (!folder.name) folder.name = 'â€¦'
  })

  return path
}

const mapStateToProps = (state, ownProps) => ({
  path: renamePathNames(
    getFolderPath(
      state.view.displayedFolder,
      state.view.currentView,
      ownProps.isPublic,
      ownProps.sharedDocuments
    ),
    ownProps.location.pathname,
    ownProps.t
  ),
  getFolderUrl
})

const mapDispatchToProps = dispatch => ({
  goToFolder: folderId => dispatch(openFolder(folderId, 'ACTION_SUCCESS'))
})

const withSharedDocuments = Wrapped =>
  class withSharedDocumentsClass extends React.Component {
    render() {
      return (
        <SharedDocuments>
          {({ sharedDocuments }) => (
            <Wrapped sharedDocuments={sharedDocuments} {...this.props} />
          )}
        </SharedDocuments>
      )
    }
  }

export default withBreakpoints()(
  withRouter(
    translate()(
      withSharedDocuments(
        connect(
          mapStateToProps,
          mapDispatchToProps
        )(MobileAwareBreadcrumb)
      )
    )
  )
)
