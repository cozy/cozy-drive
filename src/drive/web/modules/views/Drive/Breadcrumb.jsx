import React, { useCallback } from 'react'
import { useQuery, Q } from 'cozy-client'
import get from 'lodash/get'
import uniqBy from 'lodash/uniqBy'
import { useI18n } from 'cozy-ui/react/I18n'
import { MobileAwareBreadcrumbV2 as Breadcrumb } from 'drive/web/modules/navigation/Breadcrumb/MobileAwareBreadcrumb'
import { ROOT_DIR_ID } from 'drive/constants/config'

const getBreadcrumbPath = (t, displayedFolder) =>
  uniqBy(
    [
      {
        id: ROOT_DIR_ID
      },
      {
        id: get(displayedFolder, 'dir_id')
      },
      {
        id: displayedFolder.id,
        name: displayedFolder.name
      }
    ],
    'id'
  )
    .filter(({ id }) => Boolean(id))
    .map(breadcrumb => ({
      id: breadcrumb.id,
      name:
        breadcrumb.name ||
        (breadcrumb.id === ROOT_DIR_ID ? t('breadcrumb.title_drive') : 'â€¦')
    }))

const DriveBreadcrumb = ({ currentFolderId, navigateToFolder }) => {
  const { t } = useI18n()
  const currentFolderQuery = useQuery(
    () => Q('io.cozy.files').getById(currentFolderId),
    {
      as: 'folder-' + currentFolderId
    }
  )
  const currentFolder = get(currentFolderQuery, 'data[0]')
  const path = currentFolder ? getBreadcrumbPath(t, currentFolder) : []

  const onBreadcrumbClick = useCallback(({ id }) => navigateToFolder(id), [
    navigateToFolder
  ])

  return currentFolder ? (
    <Breadcrumb
      path={path}
      onBreadcrumbClick={onBreadcrumbClick}
      opening={false}
    />
  ) : null
}

export default DriveBreadcrumb
