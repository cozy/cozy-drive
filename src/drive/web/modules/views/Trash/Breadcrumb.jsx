import React from 'react'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import { MobileAwareBreadcrumbV2 as Breadcrumb } from 'drive/web/modules/navigation/Breadcrumb/MobileAwareBreadcrumb'

const RecentBreadcrumb = () => {
  const { t } = useI18n()

  return (
    <Breadcrumb
      path={[{ name: t('breadcrumb.title_trash') }]}
      opening={false}
    />
  )
}

export default RecentBreadcrumb
