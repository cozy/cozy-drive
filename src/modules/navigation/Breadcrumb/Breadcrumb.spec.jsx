import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { dummyBreadcrumbPath } from 'test/dummies/dummyBreadcrumbPath'

import Breadcrumb from './Breadcrumb'

describe('Breadcrumbs', () => {
  const path = dummyBreadcrumbPath()

  describe('template', () => {
    it('should match snapshot', () => {
      // When
      const { container } = render(<Breadcrumb path={path} />)

      // Then
      expect(container).toMatchSnapshot()
    })

    it('should be empty while path is undefined', () => {
      // When
      const { container } = render(<Breadcrumb />)

      // Then
      expect(container).toBeEmptyDOMElement()
    })

    it('should add inlined style while inlined prop true', () => {
      // When
      const { container } = render(<Breadcrumb path={path} inlined />)

      // Then
      expect(container.querySelector('.inlined')).not.toBeEmptyDOMElement()
    })
  })

  describe('events', () => {
    it('should fire on breadcrumb click when link is clicked', () => {
      // Given
      const onBreadcrumbClick = jest.fn()
      const { container } = render(
        <Breadcrumb path={path} onBreadcrumbClick={onBreadcrumbClick} />
      )

      // When
      fireEvent.click(container.querySelector('.fil-path-link'))

      // Then
      expect(onBreadcrumbClick).toHaveBeenCalledWith({ name: 'Drive' })
    })

    it('should toggle deploy on click on current', () => {
      // Given
      document.addEventListener = jest.fn()

      // Given
      const { container } = render(<Breadcrumb path={path} />)

      // When
      fireEvent.click(container.querySelector('.fil-path-current'))

      // Then
      expect(container.querySelector('.deployed')).toBeInTheDocument()
      expect(document.addEventListener).toHaveBeenCalledWith(
        'mousedown',
        expect.any(Function)
      )
    })

    it('should close menu', () => {
      // Given
      document.removeEventListener = jest.fn()
      const { container } = render(<Breadcrumb path={path} />)
      fireEvent.click(container.querySelector('.fil-path-current'))

      expect(container.querySelector('.deployed')).toBeInTheDocument()

      // When
      fireEvent.click(container.querySelector('.fil-path-current'))

      // Then
      expect(container.querySelector('.deployed')).not.toBeInTheDocument()
    })
  })
})
