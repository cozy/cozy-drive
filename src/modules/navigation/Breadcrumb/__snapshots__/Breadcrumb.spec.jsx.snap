// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs template should match snapshot 1`] = `
<div>
  <div
    class="fil-path-backdrop"
  >
    <h2
      class="fil-path-title"
      data-testid="path-title"
    >
      <span
        class="fil-path-link"
      >
        <span
          class="fil-path-link-name"
        >
          Drive
        </span>
        <svg
          class="fil-path-separator styles__icon___23x3R"
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="fil-path-link"
      >
        <span
          class="fil-path-link-name"
        >
          grandParent
        </span>
        <svg
          class="fil-path-separator styles__icon___23x3R"
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="fil-path-link"
      >
        <span
          class="fil-path-link-name"
        >
          parent
        </span>
        <svg
          class="fil-path-separator styles__icon___23x3R"
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="fil-path-current"
      >
        <span
          class="fil-path-current-name"
        >
          current
        </span>
        <span
          class="fil-path-down"
        />
      </span>
    </h2>
  </div>
</div>
`;
